File: tests/test_7_4_6_discard_step.py
"""
Tests for Task 7.4.6: discard() Step.

Validates that discard() drains the traversal while still executing
side-effect steps, and returns None (not the traversal).
See docs/project/changelog/task-7.4.6-discard-step.adoc.
"""

import pytest

from tinkercat import TinkerCat

from mocks import build_modern_graph

# ---------------------------------------------------------------------------
# Standalone SideEffectTraversal — no MockTraversal inheritance
# ---------------------------------------------------------------------------

class SideEffectTraversal:
    def __init__(self, elements):
        self._pipeline = list(elements)

    def side_effect(self, fn):
        for item in self._pipeline:
            fn(item)
        return self

    def property(self, key, value):
        for item in self._pipeline:
            item.set_property(key, value)
        return self

    def discard(self):
        self._pipeline = []
        return None

    def to_list(self):
        return list(self._pipeline)

    def has_next(self):
        return bool(self._pipeline)

# ---------------------------------------------------------------------------
# Fixtures
# ---------------------------------------------------------------------------

@pytest.fixture
def g():
    graph = TinkerCat()
    build_modern_graph(graph)
    yield graph
    graph.close()

@pytest.fixture
def src(g):
    return g.traversal()

# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_discard_returns_none(src):
    t = SideEffectTraversal(src.V().to_list())
    result = t.discard()
    assert result is None

def test_discard_on_empty_traversal_no_error(src):
    t = SideEffectTraversal([])
    result = t.discard()
    assert result is None

def test_discard_drains_pipeline(src):
    t = SideEffectTraversal(src.V().to_list())
    t.discard()
    assert t._pipeline == []

def test_side_effects_execute_before_discard(g):
    log = []
    t = SideEffectTraversal(list(g.vertices()))
    t.side_effect(lambda v: log.append(v.id))
    t.discard()
    assert len(log) == g.vertex_count

def test_property_mutation_applied_before_discard(g):
    t = SideEffectTraversal(list(g.vertices()))
    t.property("migrated", True)
    t.discard()
    for v in g.vertices():
        assert v.value("migrated") is True

def test_discard_vs_iterate_same_side_effects(g):
    log_discard = []
    log_iterate = []

    t1 = SideEffectTraversal(list(g.vertices()))
    t1.side_effect(lambda v: log_discard.append(v.id))
    t1.discard()

    t2 = SideEffectTraversal(list(g.vertices()))
    t2.side_effect(lambda v: log_iterate.append(v.id))
    t2.to_list()  # iterate-equivalent

    assert sorted(log_discard) == sorted(log_iterate)

File: tests/kotlin/test_7_4_6_discard_step.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.4.6: discard() Step.
 *
 * All tests call the discard() traversal step.  discard() throws
 * UnsupportedOperationException until implemented; property() mutation step
 * throws the same exception.
 * Note: iterate() IS implemented and is used as a comparison baseline.
 * Mirrors tests/python/test_7_4_6_discard_step.py.
 */
class Test_7_4_6_DiscardStep : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest {
        graph = TinkerCat.open()
        val marko  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "marko",  "age", 29)
        val vadas  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "vadas",  "age", 27)
        val lop    = graph.addVertex(TinkerVertex.T.label, "software", "name", "lop",    "lang", "java")
        val josh   = graph.addVertex(TinkerVertex.T.label, "person",   "name", "josh",   "age", 32)
        val ripple = graph.addVertex(TinkerVertex.T.label, "software", "name", "ripple", "lang", "java")
        val peter  = graph.addVertex(TinkerVertex.T.label, "person",   "name", "peter",  "age", 35)

        marko.addEdge("knows",   vadas,  "weight", 0.5)
        marko.addEdge("knows",   josh,   "weight", 1.0)
        marko.addEdge("created", lop,    "weight", 0.4)
        josh.addEdge("created",  ripple, "weight", 1.0)
        josh.addEdge("created",  lop,    "weight", 0.4)
        peter.addEdge("created", lop,    "weight", 0.2)
    }

    "discard returns Unit without error" {
        graph.traversal().V().discard()
        // if we reach here with Unit return, the step is implemented
    }

    "discard on empty graph does not raise" {
        val g2 = TinkerCat.open()
        try {
            g2.traversal().V().discard()
        } finally {
            g2.close()
        }
    }

    "discard drains the pipeline" {
        val t = graph.traversal().V()
        t.discard()
        t.toList() shouldBe emptyList()
    }

    "side effects execute before discard" {
        val log = mutableListOf<Any?>()
        graph.traversal().V()
            .sideEffect { v -> log.add(v.id()) }
            .discard()
        log.size shouldBe graph.vertices.size
    }

    "property mutation applied before discard" {
        graph.traversal().V()
            .property("migrated", true)
            .discard()
        for (v in graph.vertices().asSequence()) {
            v.value<Boolean>("migrated") shouldBe true
        }
    }

    "discard and iterate produce equivalent side effects" {
        val logDiscard = mutableListOf<Any?>()
        val logIterate = mutableListOf<Any?>()

        graph.traversal().V()
            .sideEffect { v -> logDiscard.add(v.id()) }
            .discard()
        graph.traversal().V()
            .sideEffect { v -> logIterate.add(v.id()) }
            .iterate()

        logDiscard.map { it.toString() }.sorted() shouldBe logIterate.map { it.toString() }.sorted()
    }
})
